### PR TITLE
at86rf2xx: reliable radio state change

### PR DESCRIPTION
**Pull request by Hyungsin**
This PR reorganizes #5971 that I did with @immesys.  
I think it was about to be merged and the reviewer was @OlegHahm. 

The following figure may be useful for review
image

### CODE REVIEW

File: drivers/at86rf2xx/at86rf2xx_internal.c [hunk 1]
```diff
@@ -127,22 +127,26 @@ void at86rf2xx_assert_awake(at86rf2xx_t *dev)
...
 {
-    /* wake up from sleep in case radio is sleeping */
-    at86rf2xx_assert_awake(dev);
-
     /* trigger hardware reset */
     gpio_clear(dev->params.reset_pin);
     xtimer_usleep(AT86RF2XX_RESET_PULSE_WIDTH);
     gpio_set(dev->params.reset_pin);
     xtimer_usleep(AT86RF2XX_RESET_DELAY);
+    dev->state = AT86RF2XX_STATE_P_ON;
```
Comment (thomaseichinger) [inline drivers/at86rf2xx/at86rf2xx_internal.c]: The state after setting the reset pin high is `AT86RF2XX_STATE_TRX_OFF`.

**Comment by Hyungsin** [comment]
@thomaseichinger, I updated regarding your review :)

**Comment by smlng** [comment]
> This PR reorganizes #5971

so the other one can be closed, or does/should this depend on the other?

**Comment by Hyungsin** [comment]
@smlng, @immesys, I think #5971 is good to be closed.
This PR incorporates #5971 :)

**Comment by smlng** [comment]
@Hyungsin then close it please, if @immesys is fine with that, too.

Further, regarding this PR here: though less commits are good, IMHO here 2 separate would be better: one for the changed timings (internal header) and another for the refined handling.

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 2]
```diff
@@ -440,6 +440,12 @@ static inline void _set_state(at86rf2xx_t *dev, uint8_t state, uint8_t cmd)
     if (state != AT86RF2XX_STATE_RX_AACK_ON) {
         while (at86rf2xx_get_status(dev) != state) {}
     }
+    /* Although RX_AACK_ON state doesn't get read back, 
+     * at least make sure if state transition is in progress or not 
+     */ 
+    else {
+        while (at86rf2xx_get_status(dev) == AT86RF2XX_STATE_IN_PROGRESS);
```
Comment (smlng) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: coding style: use `{}` instead of `;`

File: drivers/at86rf2xx/at86rf2xx.c [hunk 3]
```diff
@@ -44,7 +44,7 @@ void at86rf2xx_setup(at86rf2xx_t *dev, const at86rf2xx_params_t *params)
     /* initialize device descriptor */
     memcpy(&dev->params, params, sizeof(at86rf2xx_params_t));
     dev->idle_state = AT86RF2XX_STATE_TRX_OFF;
-    dev->state = AT86RF2XX_STATE_SLEEP;
+    dev->state = AT86RF2XX_STATE_P_ON;  /* radio state is P_ON when first powered-on */
```
Comment (smlng) [inline drivers/at86rf2xx/at86rf2xx.c]: put comment above to reduce line length (80 chars is encourage, where possible)

File: drivers/at86rf2xx/at86rf2xx_internal.c [hunk 4]
```diff
@@ -127,22 +127,34 @@ void at86rf2xx_assert_awake(at86rf2xx_t *dev)
...
-        dev->state = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_STATUS)
-                         & AT86RF2XX_TRX_STATUS_MASK__TRX_STATUS;
+        /* update state: on some platforms, the timer behind xtimer 
+         * may be inaccurate or the radio itself may take longer 
+         * to wake up due to extra capacitance on the oscillator. 
+         * Spin until we are actually awake
+         */
+        do {
+          dev->state = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_STATUS)
+                           & AT86RF2XX_TRX_STATUS_MASK__TRX_STATUS;
```
Comment (smlng) [inline drivers/at86rf2xx/at86rf2xx_internal.c]: please align, like:

```
    dev->state = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_STATUS) &
                 AT86RF2XX_TRX_STATUS_MASK__TRX_STATUS;
```

File: drivers/at86rf2xx/at86rf2xx_internal.c [hunk 5]
```diff
@@ -127,22 +127,34 @@ void at86rf2xx_assert_awake(at86rf2xx_t *dev)
...
     gpio_clear(dev->params.reset_pin);
     xtimer_usleep(AT86RF2XX_RESET_PULSE_WIDTH);
     gpio_set(dev->params.reset_pin);
     xtimer_usleep(AT86RF2XX_RESET_DELAY);
+
+    /* update state: if the radio state was P_ON (initialization phase), 
+     * it remains P_ON. Otherwise, it should go to TRX_OFF 
+     */
+    do {
+        dev->state = at86rf2xx_reg_read(dev, AT86RF2XX_REG__TRX_STATUS)
```
Comment (smlng) [inline drivers/at86rf2xx/at86rf2xx_internal.c]: indention/alignment, see above

**Comment by Hyungsin** [comment]
@smlng, That makes sense. 
Then, I will move "at86rf2xx_internal.h" to another PR.

File: drivers/at86rf2xx/at86rf2xx_netdev.c [hunk 6]
```diff
@@ -77,8 +77,8 @@ static int _init(netdev_t *netdev)
     gpio_set(dev->params.reset_pin);
     gpio_init_int(dev->params.int_pin, GPIO_IN, GPIO_RISING, _irq_handler, dev);
 
-    /* make sure device is not sleeping, so we can query part number */
-    at86rf2xx_assert_awake(dev);
```
Comment (smlng) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: (question): I thought for some settings the device must be powered up, so it safe to remove this here?

**Comment by smlng** [comment]
> Then, I will move "at86rf2xx_internal.h" to another PR.

it seems a bit unrelated here, yes - but a separate commit in this PR would also be fine.

**Comment by Hyungsin** [comment]
Fixed style and separated commits :)

File: drivers/at86rf2xx/at86rf2xx_netdev.c (hunk 6 above)
Comment (thomaseichinger) [inline drivers/at86rf2xx/at86rf2xx_netdev.c]: @smlng `at86rf2xx_reset` calls `at86rf2xx_hardware_reset` which results in either `P_ON` or `TRX_OFF` states, depending of the state before.

**Comment by thomaseichinger** [comment]
@Hyungsin Please consider https://github.com/Hyungsin/RIOT-OS/pull/2 for piggypacking some more cleanup and reduce code size (at least a little bit).

Also please prefix your commit messages with `drivers/at86rf2xx: `.

**Comment by Hyungsin** [comment]
@thomaseichinger, Done :)

**Comment by Hyungsin** [comment]
@thomaseichinger, @smlng, any other comments? Good to go?

**Comment by smlng** [comment]
I'm fine with the changes, though we should be careful with changes in timing - but as far as I see the timing are related to reset and wake up, so we should be fine here. ACK!

**Comment by Hyungsin** [comment]
@smlng, Thanks! and @thomaseichinger ?

**Comment by Hyungsin** [comment]
Please check @thomaseichinger :)

**Comment by thomaseichinger** [comment]
@Hyungsin Sorry for the delay, started CI build.

**Comment by Hyungsin** [comment]
@thomaseichinger, What should I fix? 
I don't know much about CI building process :(

**Comment by immesys** [comment]
@Hyungsin it's just whitespace stuff. You can see in the build output where the cause is

**Comment by Hyungsin** [comment]
Oh, all checks have passed :) Thank you guys.

**Comment by thomaseichinger** [comment]
Awesome, ACK & go.

**Pull request merged**